avalanche/training/templates/observation_type/batch_observation.py

```diff
@@ -0,0 +1,46 @@
+from typing import Iterable
+
+from avalanche.benchmarks import CLExperience
+from avalanche.models.dynamic_optimizers import reset_optimizer
+
+
+class BatchObservation:
+    def _train_exp(
+        self, experience: CLExperience, eval_streams=None, **kwargs
+    ):
+        """Training loop over a single Experience object.
+
+        :param experience: CL experience information.
+        :param eval_streams: list of streams for evaluation.
+            If None: use the training experience for evaluation.
+            Use [] if you do not want to evaluate during training.
+        :param kwargs: custom arguments.
+        """
+        if eval_streams is None:
```

Comment:
L19-L26 should be factored out of this method since it's common for all strategies

Reply:
The loop over `training_epochs` and the plugin triggers can be different in different observations, that's why I explicitly defined them here.

Reply:
Do we have a clear definition of template/observation/problem/update? For example a clear interface that they must specify